/*
 * Copyright 2018- The Pixie Authors.
 * Modifications Copyright 2023- Gimlet Labs, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

syntax = "proto3";

package gml.statuspb;

option go_package = "statuspb";

import "google/protobuf/any.proto";

enum Code {
  // buf:lint:ignore ENUM_ZERO_VALUE_SUFFIX
  // Not an error, returned on success.
  CODE_OK = 0;
  // The operation was cancelled (typically by the caller).
  CODE_CANCELLED = 1;
  // There was an error due to unknown cause. Errors may be converted
  // into this form if APIs don't give enough information to get the correct
  // error codes.
  CODE_UNKNOWN = 2;
  // Bad argument was given and a retry will not be successful.
  CODE_INVALID_ARGUMENT = 3;
  // The deadline for this operation was exceeded.
  CODE_DEADLINE_EXCEEDED = 4;
  // Resource specified was not found.
  CODE_NOT_FOUND = 5;
  // The resource already exists and cannot be overridden.
  CODE_ALREADY_EXISTS = 6;
  CODE_PERMISSION_DENIED = 7;
  CODE_UNAUTHENTICATED = 8;
  CODE_INTERNAL = 9;
  // Operation specified is not implemented.
  CODE_UNIMPLEMENTED = 10;
  // No resources available for action.
  CODE_RESOURCE_UNAVAILABLE = 11;
  // Syscall errors.
  CODE_SYSTEM = 12;
  // A precondition of the API failed. Might have succeeded otherwise.
  CODE_FAILED_PRECONDITION = 13;
  // Use default instead of specifying all entries here. Otherwise, code might get
  // hard to maintain.
  CODE_DO_NOT_USE = 100;
}

// This is the proto representation of a status.
message Status {
  Code err_code = 1;
  string msg = 2;
  google.protobuf.Any context = 3;
}
